### ISSUE TITLE
Engines for H2O tree models

### ISSUE THREAD

**Issue by riccardopinosio** [tidymodels/parsnip#205]
I have just started experimenting with parsnip and I think the package is awesome. It's a really needed package as well as it's high time R has something that can potentially rival scikit in python.
I would really like to contribute to the package as I am currently doing a fair amount of modelling work.

A first thing I could do is contribute code to add H2O as an engine to e.g. the random forest and xgboost models (to start with). Would this be a welcome contribution? And, is there a contribution guideline?

**Comment by topepo** [comment]
@ledell started a fork to do this but I was in the process of re-writing how to register models in `parsnip`. 

We are definitely doing this and you are welcome to pitch in. Otherwise, it'll probably be a month or so before I approach it.

**Comment by ledell** [comment]
@riccardopinosio It would be great if you help @topepo add this.  I am available to help too.

**Comment by riccardopinosio** [comment]
Ok, I will start looking at it this week.
@ledell is there code in your fork that can still be used, or given that the registration of models changed it's better to start from scratch?

**Comment by ledell** [comment]
@riccardopinosio I think it needs to be written from scratch...

**Comment by ZarniHtet13** [comment]
Hi All, 

Just curious to the timeline of when h2o powered models would become available on parsnip.
Thanks a lot!

ZH

**Comment by riccardopinosio** [comment]
@ZarniHtet13 
I haven't had time to work on it of late, I don't think I will have time in the next few weeks, so if anyone wants to take care of this do go ahead :)

**Comment by ZarniHtet13** [comment]
@riccardopinosio , thanks for getting back. I don't think I am capable of doing that. Just want to use parsnip with h2o at some point! Looking forward to it from someone :)

**Comment by Rcubes** [comment]
What kind of task are needed. It would be like registering the models following the tutorials in the parsnip site or something more complex than that?

**Comment by kdpsingh** [comment]
I would be happy to get this party started! Will fork and share updates soon. Have been using h2o for a while (both directly and via mlr) but would like to migrate to parsnip.

**Comment by stevenpawley** [comment]
Hello,

FYI I have a few h2o algorithms implemented within the parsnip API, compromising random forests, neural networks and gbm:

https://github.com/stevenpawley/h2oparsnip

I was aiming to add GLM, SVM and naive Bayes over the next couple of days.

**Comment by kdpsingh** [comment]
@stevenpawley it would make sense to merge these into the main parsnip package if that is within the scope (which is sounds like it is from @topepo’s response above).

Would you like to move some of your package code into the parsnip package itself? Or would you rather someone else (e.g., me) move it over?

**Comment by stevenpawley** [comment]
Happy for it to go either way. At one point @topepo was intending to minimize the number of wrappers contained within parsnip. h2o appears to require wrappers because the models don't employ the traditional formula or data.frame/matrix interface, and require more than one function call (to coerce into an H2OFrame, and then training). Not sure where this stands now.

**Comment by stevenpawley** [comment]
Thanks, that's a good suggestion in terms of outlining that. I'll add something details onto the h2oparsnip package shortly.

**Comment by ledell** [comment]
@stevenpawley Cool to hear about the h2oparsnsip package. :-)  I think that RStudio was hoping to have native H2O support at some point -- the most complex part is how to efficiently manage the data conversions between data.frame and H2OFrames.  I don't think there's been any progress on that yet, but perhaps they can re-use some of your work in h2oparsnip when the integration happens.

**Comment by stevenpawley** [comment]
Thanks! At the moment, the 'h2oparsnip' package is mostly just some wrappers to add h2o models as engines. Not sure if this will ever make it to being a standalone package and happy if parts of it a consumed into something else.

In terms of data conversions between data.frame and H2OFrames, I'm mainly selfishly working to satisfy my own requirements where I don't want all of my data in the cluster, but rather use it as a disposable backend. Also, if I have chosen to use tidymodels, then I am also likely to be using some recipe steps as part of the workflow, and thus I'm accepting that there will have to be some conversions for calculations that occur outside of the h2o cluster, otherwise, I'd just be using h2o directly.

That said, I would like to mitigate some of the major bottlenecks caused by lots of data going back and forth during tuning and cross-validation. I have a rough version of function `tune_grid_h2o` that is an (almost) drop-in replacement for `tune::tune_grid` but it uses `h2o.Grid` to perform hyperparameter tuning of model arguments all within the cluster. However, two differences are that (a) only model arguments can be tuned (not recipes); and (b) AFAIU handling of metrics has to be different because in `tune_grid` you can use any yardstick metric or a custom metric to tune against, but `h2o.Grid` requires one of the specified metrics (e.g. mse, accuracy, error, mean_per_class_error etc). It wouldn't be hard to wrap `tune_grid_h2o` inside something else that calls it repeatedly to tune recipe arguments with the tuning results being merged together. Otherwise, `tune_grid_h2o` returns the tuning results as a rsample object that can be used with all of the regular tidymodels functions, e.g. `select_best` etc. Another difference is AFAIK, `h2o.Grid` does not store the predictions during tuning, whereas `tune::tune_grid` gives you the option to extract the predictions via `control_grid()`.

Diverging more from tidymodels, I was also thinking of a `cross_validate_h2o` function as a generic to take either a `rset` object for non-nested cross-validation or a `nested_cv` object to perform nested cross-validation. This would mitigate the second bottle-neck by using h2o to perform most of the operations in the cluster, while still returning rsample-like objects for analysis in tidymodels style. Not sure if this is moving in the right direction for everyone, especially for users who want to use h2o for optimal performance with all data/processing occurring in H2OFrames, but the other tidymodels packages, e.g. rsample, recipes etc. currently only accept data.frames, so again I'm accepting that some conversions must occur.

**Comment by simonpcouch** [comment]
Closing in light of development on [the agua package](https://github.com/tidymodels/agua)! Thanks all for contributions and input. :)
